qiskit/primitives/backend_estimator.py

```diff
@@ -186,30 +186,40 @@ def _transpile(self):
         self._transpiled_circuits = []
         for common_circuit, diff_circuits in self.preprocessed_circuits:
             # 1. transpile a common circuit
-            common_circuit = common_circuit.copy()
-            num_qubits = common_circuit.num_qubits
-            common_circuit.measure_all()
             if not self._skip_transpilation:
-                common_circuit = transpile(
+                transpiled_circuit = transpile(
                     common_circuit, self.backend, **self.transpile_options.__dict__
                 )
-            bit_map = {bit: index for index, bit in enumerate(common_circuit.qubits)}
-            layout = [bit_map[qr[0]] for _, qr, _ in common_circuit[-num_qubits:]]
-            common_circuit.remove_final_measurements()
+                if transpiled_circuit.layout is not None:
+                    layout = transpiled_circuit.layout
+                    perm_pattern = [
+                        layout.initial_layout.get_virtual_bits()[v] for v in common_circuit.qubits
+                    ]
```

Comment:
This will have a non-zero overhead which you probably want to avoid in a list comprehension. I would bring the layout call outside the comprehension like:

```suggestion
                    virtual_bit_map = layout.initial_layout.get_virtual_bits()
                    perm_pattern = [
                        virtual_bit_map[v] for v in common_circuit.qubits
                    ]
```

Reply:
Thank you very much. Indeed it is.